Deprecate calling Task methods at execution time (#37551)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_9.adoc

implementation.getDependencies().add(dependency);
----

[[task_dependencies]]
==== Accessing task dependency relationships from a task action

Calling the following methods from a task action at execution time is now deprecated and will become an error in Gradle 10:

- link:{javadocPath}/org/gradle/api/Task.html#getTaskDependencies--[`Task.getTaskDependencies()`]
- link:{javadocPath}/org/gradle/api/Task.html#getDependsOn--[`Task.getDependsOn()`]
- link:{javadocPath}/org/gradle/api/Task.html#getMustRunAfter--[`Task.getMustRunAfter()`]
- link:{javadocPath}/org/gradle/api/Task.html#getFinalizedBy--[`Task.getFinalizedBy()`]
- link:{javadocPath}/org/gradle/api/Task.html#getShouldRunAfter--[`Task.getShouldRunAfter()`]

These methods can still be used during configuration time.

The deprecation is only issued if the Configuration Cache is **not** enabled.
When the Configuration Cache is enabled, calls to these methods are reported as Configuration Cache problems instead.

> The deprecation of `Task.getTaskDependencies()` was originally introduced in <<upgrading_version_7.adoc#task_dependencies, Gradle 7.4>> but was only issued when the <<configuration_cache_enabling.adoc#config_cache:stable,`STABLE_CONFIGURATION_CACHE`>> feature flag was enabled.

Reviewer: 🤔 The history is a bit off here. I'm not sure if we need to lay out all the changes, but...

- `Task.getTaskDependencies` and `Task.getDependsOn` became hard errors under `STABLE_CONFIGURATION_CACHE` in 8.0. They were deprecations since 7.4.
-  `getMustRunAfter`, `getFinalizedBy`, and `getShouldRunAfter` joined the club in 9.0.

implementation.getDependencies().add(dependency);
----

[[task_dependencies]]
==== Accessing task dependency relationships from a task action

> Calling the following methods from a task action at execution time is now deprecated and will be made an error in Gradle 10:

Reviewer 2: ```suggestion
Calling the following methods from a task action at execution time is now deprecated and will become an error in Gradle 10:
```

> - link:{javadocPath}/org/gradle/api/Task.html#getTaskDependencies--[Task.getTaskDependencies()]

Reviewer 2: ```suggestion
- link:{javadocPath}/org/gradle/api/Task.html#getTaskDependencies--[`Task.getTaskDependencies()`]
```

> - link:{javadocPath}/org/gradle/api/Task.html#getDependsOn--[Task.getDependsOn()]

Reviewer 2: ```suggestion
- link:{javadocPath}/org/gradle/api/Task.html#getDependsOn--[`Task.getDependsOn()`]
```

> - link:{javadocPath}/org/gradle/api/Task.html#getMustRunAfter--[Task.getMustRunAfter()]

Reviewer 2: ```suggestion
- link:{javadocPath}/org/gradle/api/Task.html#getMustRunAfter--[`Task.getMustRunAfter()`]
```

> - link:{javadocPath}/org/gradle/api/Task.html#getFinalizedBy--[Task.getFinalizedBy()]

Reviewer 2: ```suggestion
- link:{javadocPath}/org/gradle/api/Task.html#getFinalizedBy--[`Task.getFinalizedBy()`]
```

> - link:{javadocPath}/org/gradle/api/Task.html#getShouldRunAfter--[Task.getShouldRunAfter()]

Reviewer 2: ```suggestion
- link:{javadocPath}/org/gradle/api/Task.html#getShouldRunAfter--[`Task.getShouldRunAfter()`]
```

These methods can still be used during configuration time.

> The deprecation is only issued if the configuration cache is **not** enabled.

Reviewer 2: ```suggestion
The deprecation is only issued if the Configuration Cache is **not** enabled.
```

> When the configuration cache is enabled, calls to these methods are reported as configuration cache problems instead.

Reviewer 2: ```suggestion
When the Configuration Cache is enabled, calls to these methods are reported as Configuration Cache problems instead.
```

The deprecation of `Task.getTaskDependencies()` was originally introduced in <<upgrading_version_7.adoc#task_dependencies, Gradle 7.4>> but was only issued when the <<configuration_cache_enabling.adoc#config_cache:stable,`STABLE_CONFIGURATION_CACHE`>> feature flag was enabled.
That feature flag no longer controls this deprecation.
The other methods listed above are newly deprecated in this version.

> This is another step towards moving users away from idioms that are incompatible with the configuration cache, which will become the only mode supported by Gradle in a future release.

Reviewer 2: ```suggestion
This is another step towards moving users away from idioms that are incompatible with the Configuration Cache, which will become the only mode supported by Gradle in a future release.
```

> 

Reviewer 2: It would be nice to at least add one code example (please update the example as needed):
```suggestion
For example:

[source,kotlin]
----
// Deprecated: inspecting dependencies from within the task action
tasks.register("report") {
    dependsOn("compileJava")
    doLast {
        taskDependencies.getDependencies(this).forEach { println(it.name) }
    }
}

// Recommended: capture what you need at configuration time
tasks.register("report") {
    dependsOn("compileJava")
    val dependencyNames = dependsOn.filterIsInstance<Task>().map { it.name }
    doLast {
        dependencyNames.forEach { println(it) }
    }
}
----

```

> Please refer to the <<configuration_cache_requirements.adoc#config_cache:requirements:use_project_during_execution, configuration cache documentation>> for alternatives that are compatible with the configuration cache.

Reviewer 2: ```suggestion
Please refer to the <<configuration_cache_requirements.adoc#config_cache:requirements:use_project_during_execution, Configuration Cache documentation>> for alternatives that are compatible with the Configuration Cache.
```

[[task_extensions]]
==== Accessing extensions from a task action

> Calling link:{javadocPath}/org/gradle/api/Task.html#getExtensions--[Task.getExtensions()] from a task action at execution time is now deprecated and will be made an error in Gradle 10.

Reviewer 2: ```suggestion
Calling link:{javadocPath}/org/gradle/api/Task.html#getExtensions--[`Task.getExtensions()`] from a task action at execution time is now deprecated and will become an error in Gradle 10.
```

This method can still be used during configuration time.

> The deprecation is only issued if the configuration cache is **not** enabled.

Reviewer 2: ```suggestion
The deprecation is only issued if the Configuration Cache is **not** enabled.
```

> When the configuration cache is enabled, calls to this method are reported as configuration cache problems instead.

Reviewer 2: ```suggestion
When the Configuration Cache is enabled, calls to this method are reported as Configuration Cache problems instead.
```

This deprecation was originally introduced in <<upgrading_version_8.adoc#task_convention, Gradle 8.2>> but was only issued when the <<configuration_cache_enabling.adoc#config_cache:stable,`STABLE_CONFIGURATION_CACHE`>> feature flag was enabled.
That feature flag no longer controls this deprecation.

> This is another step towards moving users away from idioms that are incompatible with the configuration cache, which will become the only mode supported by Gradle in a future release.

Reviewer 2: ```suggestion
This is another step towards moving users away from idioms that are incompatible with the Configuration Cache, which will become the only mode supported by Gradle in a future release.
```

> 

Reviewer 2: ```suggestion

For example:
[source,kotlin]
----
// Deprecated: reading the extension from within the task action
tasks.register("greet") {
    doLast {
        val ext = extensions.getByType<GreetingExtension>()
        println(ext.message.get())
    }
}

// Recommended: capture the value at configuration time
tasks.register("greet") {
    val message = extensions.getByType<GreetingExtension>().message
    doLast {
        println(message.get())
    }
}
----

```

> Please refer to the <<configuration_cache_requirements.adoc#config_cache:requirements:use_project_during_execution, configuration cache documentation>> for alternatives that are compatible with the configuration cache.

Reviewer 2: ```suggestion
Please refer to the <<configuration_cache_requirements.adoc#config_cache:requirements:use_project_during_execution, Configuration Cache documentation>> for alternatives that are compatible with the configuration cache.
```

[[injected_service_types_at_execution]]
==== Accessing injected `Project` or `Gradle` services from a task action

> Reading an injected service of type link:{javadocPath}/org/gradle/api/Project.html[Project] or link:{javadocPath}/org/gradle/api/invocation/Gradle.html[Gradle] from a task action at execution time is now deprecated and will be made an error in Gradle 10.

Reviewer 2: ```suggestion
Reading an injected service of type link:{javadocPath}/org/gradle/api/Project.html[`Project`] or link:{javadocPath}/org/gradle/api/invocation/Gradle.html[`Gradle`] from a task action at execution time is now deprecated and will become an error in Gradle 10.
```

These services can still be read during configuration time.

> The deprecation is only issued if the configuration cache is **not** enabled.

Reviewer 2: ```suggestion
The deprecation is only issued if the Configuration Cache is **not** enabled.
```

> When the configuration cache is enabled, such accesses are reported as configuration cache problems instead.

Reviewer 2: ```suggestion
When the Configuration Cache is enabled, such accesses are reported as Configuration Cache problems instead.
```

This restriction was originally enforced only when the <<configuration_cache_enabling.adoc#config_cache:stable,`STABLE_CONFIGURATION_CACHE`>> feature flag was enabled.
That feature flag no longer controls this deprecation.

> This is another step towards moving users away from idioms that are incompatible with the configuration cache, which will become the only mode supported by Gradle in a future release.

Reviewer 2: ```suggestion
This is another step towards moving users away from idioms that are incompatible with the Configuration Cache, which will become the only mode supported by Gradle in a future release.
```

> 

Reviewer 2: ```suggestion

For example:

[source,kotlin]
----
// Deprecated: reading from the injected Project at execution time
abstract class PrintVersionTask : DefaultTask() {
    @get:Inject
    abstract val project: Project

    @TaskAction
    fun run() {
        println("Building ${project.name} version ${project.version}")
    }
}

// Recommended: declare the values you need as task inputs
abstract class PrintVersionTask : DefaultTask() {
    @get:Input
    abstract val projectName: Property<String>

    @get:Input
    abstract val projectVersion: Property<String>

    @TaskAction
    fun run() {
        println("Building ${projectName.get()} version ${projectVersion.get()}")
    }
}

tasks.register<PrintVersionTask>("printVersion") {
    projectName.set(project.name)
    projectVersion.set(project.version.toString())
}
----
```

> Please refer to the <<configuration_cache_requirements.adoc#config_cache:requirements:use_project_during_execution, configuration cache documentation>> for alternatives that are compatible with the configuration cache.

Reviewer 2: ```suggestion
Please refer to the <<configuration_cache_requirements.adoc#config_cache:requirements:use_project_during_execution, Configuration Cache documentation>> for alternatives that are compatible with the Configuration Cache.
```
